File: problem/Min_Operations_Equalize/minOperations.c
#include <string.h>
#include <math.h>
#include <limits.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))
#define MIN(a, b) ((a) < (b) ? (a) : (b))

int minOperations(char *s, int k)
{
    int zero = 0;
    int len = strlen(s);

    for (int i = 0; i < len; i++)
        zero += ~s[i] & 1;

    if (!zero)
        return 0;

    if (len == k)
        return ((zero == len) << 1) - 1;

    int base = len - k;

    int odd = MAX(
        (zero + k - 1) / k,
        (len - zero + base - 1) / base);
    odd += ~odd & 1;

    int even = MAX(
        (zero + k - 1) / k,
        (zero + base - 1) / base);
    even += even & 1;

    int res = INT_MAX;

    if ((k & 1) == (zero & 1))
        res = MIN(res, odd);

    if (~zero & 1)
        res = MIN(res, even);

    return res == INT_MAX ? -1 : res;
}

File: problem/Min_Operations_Equalize/minOperations.py
class Solution:
    def minOperations(self, s: str, k: int) -> int:
        zero = 0
        length = len(s)

        for char in s:
            zero += ~ord(char) & 1

        if not zero:
            return 0

        if length == k:
            return ((zero == length) << 1) - 1

        base = length - k

        odd = max(
            (zero + k - 1) // k,
            (length - zero + base - 1) // base
        )
        odd += ~odd & 1

        even = max(
            (zero + k - 1) // k,
            (zero + base - 1) // base
        )
        even += even & 1

        res = float('inf')

        if (k & 1) == (zero & 1):
            res = min(res, odd)

        if ~zero & 1:
            res = min(res, even)

        return int(res) if res != float('inf') else -1